### ISSUE TITLE
Detect assignment to readonly variable

### ISSUE THREAD

**Issue by Warbo** [koalaman/shellcheck#1287]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

I was writing a script which involved a "unique ID", and foolishly assigned it to a variable called `UID`. This causes a bash error when run: `UID: readonly variable`. I run all my scripts through shellcheck, but it didn't catch this. It would be nice if it could.

#### Here's a snippet or screenshot that shows the problem:

The snippet `UID="anything"` should trigger this. Here's the whole script I wrote, for context:

```sh
#!/usr/bin/env bash
set -e

# Open PDFs in mupdf

function isOpen {
    # shellcheck disable=SC2009
    ps auxww | grep mupdf-x11 | grep "$1" > /dev/null
}

function openIfNeeded {
    # Open $2 if it's not already (i.e. if `mupdf-x11 ... $1` appears in `ps`)
    if isOpen "$1"
    then
        echo "Already have reader for $2, skipping" 1>&2
    else
        echo "Opening $2" 1>&2
        mupdf-x11 "$2" &
    fi
}

while read -r LINE
do
    # Skip comments
    echo "$LINE" | grep '^\s*#' > /dev/null && continue
    UID=$(echo "$LINE" | cut -f1)
    LOC=$(echo "$LINE" | cut -f2)
    openIfNeeded "$UID" "$LOC"
done < ~/.pdfs
```

Running a script like this gives:

```sh
$ pdfs
/nix/store/87lwk6i1hjfmswafhvbwsdh9m8rgjamj-warbo-utilities/bin/.pdfs-wrapped: line 26: UID: readonly variable
```

#### Here's what shellcheck currently says:

```
No issues detected!
```

#### Here's what I wanted or expected to see:

Something like:

```
The variable "UID" is readonly and built in to the shell. Trying to assign it to a new value will cause an error.
```

Other readonly variables (built in and perhaps even user-defined) could be spotted too.

**Comment by nkakouros** [comment]
Custom readonly variables created with `declare -r` could also be included in the check.

**Comment by bendavies** [comment]
and things like this:
```
function foo {
  local -r var="foo"
  var="bar"
}
```

**Comment by Keisial** [comment]
bash default variables which are readonly: `BASHOPTS`, `BASH_REMATCH`, `BASH_VERSINFO`, `EUID`, `PPID`, `SHELLOPTS`, `UID`.

**Comment by wyardley** [comment]
another odd thing.. with the code:
```shell
readonly DELETE_AT_EXIT="$(mktemp -d)"
```
I get the warning:
```
    readonly DELETE_AT_EXIT="$(mktemp -d)"
             ^------------^ SC2155 (warning): Declare and assign separately to avoid masking return values.
```
(I don't think it's possible to assign separately in this case, right, since if we declare it as readonly, we can't then assign anything to it?)

**Comment by smolin** [comment]
Also GROUPS, SECONDS, RANDOM, LINENO and FUNCNAME

This might help: https://tldp.org/LDP/abs/html/internalvariables.html

**Comment by sandorex** [comment]
Note that `GROUPS`, `RANDOM`, `LINENO`, `FUNCNAME` are not writable but also **NOT AN ERROR** when you write to it

While `SECONDS` you can write to (to reset it for example) but it will start incrementing
